untrusted/urts/src/lib.rs

```diff
@@ -19,19 +45,51 @@ pub enum Error {
 /// Avoid storing the de-referenced instance of the enclave.  The de-referenced
 /// value of the enclave will result in failures to the SGX SDK after the
 /// enclave is dropped.
-#[derive(Debug, PartialEq)]
 pub struct Enclave {
     // The enclave ID, assigned by the SGX interface
     id: sgx_enclave_id_t,
+    report_fn: Option<EnclaveReportFn>,
+}
+
+impl Enclave {
+    //TODO should this be called "get_report"?  Named create to match the sgx interface?
+    /// Return report for this enclave.
+    ///
+    /// # Arguments
+    /// - `target_info` The target info to use in creating the report.
+    ///     See [EnclaveReportFn]
+    ///
+    /// # Errors
+    /// - `[Error::NoReportFunction]` when this instance was created without a
+    ///     `[EnclaveReportFn]`, see [`EnclaveBuilder::report_fn()`] for setting
+    ///     the function.
+    /// - `[Error::SgxStatus]` when there is any error communicating to the
+    ///     SGX SDK.
+    pub fn create_report(
+        &self,
+        target_info: Option<&sgx_target_info_t>,
+    ) -> Result<sgx_report_t, Error> {
+        match self.report_fn {
+            None => Err(Error::NoReportFunction),
+            Some(report_fn) => report_fn(self, target_info),
+        }
+    }
+}
+
+impl fmt::Debug for Enclave {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "id: {}", self.id)
+    }
 }
 
 /// Build an [Enclave] for use with SGX calls.
 pub struct EnclaveBuilder {
     // The bytes for the enclave.
     bytes: Vec<u8>,
-
     // `true` if the enclave should be created in debug mode
     debug: bool,
+    // The report function to populate the [Enclave] with.  Defaults to `None`.
+    report_fn: Option<EnclaveReportFn>,
```

Comment:
I'm not convinced this should be optional. It seems better to provide a sensible default, and allow overriding it.

Reply:
Do you have thoughts on how to provide a default?
I wasn't sure the best place to put it but, https://github.com/mobilecoinfoundation/sgx/blob/feature/enclave_report/untrusted/urts/src/lib.rs#L13
```rust
/// While there is an interface to get an enclave report from inside the
/// enclave, `sgx_create_report()`, there is no standard way to get the
/// report from the untrusted side.
///
/// The [EnclaveReportFn] allows the users of [Enclave]s a general interface
/// to redirect to the specific ecall that the enclave may use for providing
/// back the report.

```
There isn't a common ecall to create the report.  Once in the enclave one can call `sgx_create_enclave()`, but it's up to enclave authors how the report itself is requested through the ecall interface

Reply:
I see. Leaving it optional SGTM then.